ludwig/hyperopt/execution.py

```diff
@@ -844,15 +898,33 @@ def get_parameters_dict(parameters):
     return parameters_dict
 
 
-def substitute_parameters(config, parameters):
+def substitute_parameters(config, parameters, shared_params_features_dict):
```

Comment:
@justinxzhao will add type hints! 

Theoretically speaking, `shared_params_features_dict` can never be of type `None` with the implementation I have, but the reason I have allowed it to be None is that there are quite a few hyperopt tests that directly use the `HyperoptExecutor.execute()` function instead of calling the top level `hyperopt()` function, which overrides the creation of `shared_params_features_dict` that happens with the top level `hyperopt()` function. 

One option is to modify all the tests that display this behavior, or to just state in the type hints that it can be None. 

Should I just modify all of the tests so that it's done correctly via the top level function?

Reply:
I would probably not modify all of the tests since it's easy enough to handle the `None` case.

One additional suggestion is to pass `shared_params_features_dict` as is, and then do shared_params_features_dict[shared_params_type] within `substitute_parameters()`.

```
shared_params_type=INPUT_FEATURES,
shared_params_features_dict=shared_params_features_dict),
```